### PR TITLE
LDG-10: Image MINV

### PR DESCRIPTION
**Pull request by fazelehh**
# Description

Clean up the image-based PLGMI model inversion attack and introduce a modality abstraction layer. This PR contains only the reliable, tested image/CelebA MINV code — tabular extensions are deferred to a separate PR.

* Add `modality_extension.py` abstract base class for modality-specific extensions
* Refactor `image_extension.py` from 194 to ~35 lines using the new abstraction
* Clean `plgmi.py` to image-only path: remove all tabular/dataframe branches, CTGAN loading, and leftover WIP/debug methods (`optimize_z_grad_original`, `optimize_z_grad2`, `load_reference_model`, etc.)
* Remove hardcoded checkpoint paths (`gen_3000.pth`/`dis_3000.pth`) from `celebA_plgmi_handler.py`
* Fix `audit.yaml` `data_modality` field (was commented out)
* Add optional discriminator support in `gan_handler.py`
* Refactor `max_margin_loss` in `gan_losses.py`
* Add `pytorch_tabular` to allowed `model_type` values in `schemas.py`
* Apply ruff linting fixes across all changed files

## Resolved Issues

- [ ] fixes #Issue

## How Has This Been Tested?

Manually tested end-to-end on the CelebA example (`examples/minv/celebA/`) with a trained ResNet152 target model.

## Related Pull Requests

- Splits off from #402 (image-only subset — tabular work deferred to follow-up PRs)

### CODE REVIEW

File: leakpro/input_handler/modality_extensions/image_extension.py [hunk 1]
```diff
@@ -1,194 +1,35 @@
...
-            self.mean = None
-            self.std = None
-        self.force_rgb = force_rgb
-        self._reg = _build_registry_mild()
-        self._tiers = _policy_tiers()
-        self.set_augment_strength(augment_strength)
+        super().__init__(handler)
+        logger.info("Image extension initialized.")
 
-    def set_augment_strength(self, augment_strength: str) -> None:
```
Comment (henrikfo) [inline leakpro/input_handler/modality_extensions/image_extension.py:104]: Does all attacks still work with this function gone from the image extension? From what I can see, RaMIA.py, line 75-76, will probably fail since "set_augment_strenght()" is removed but still called in ramia.

File: leakpro/input_handler/modality_extensions/image_extension.py
Comment (henrikfo) [inline leakpro/input_handler/modality_extensions/image_extension.py:1]: The refactor looks clean and great! But the functions removed, are they outdated or do they exists somewhere else? Are we dropping them

File: leakpro/input_handler/modality_extensions/image_extension.py [hunk 2]
```diff
@@ -1,194 +1,35 @@
...
-        tf = self._augment(num_ops=num_ops)
-        ys, ops_used = [], []
-        for j in range(k):
-            _seed = base_seed + j
-            y, ops = tf(x, _seed)
-            ys.append(y)
-            ops_used.append(ops)
-        out = torch.stack(ys, dim=0) if stack else ys
-        return out, ops_used
+        n_aug = n_aug // len(data) #dummy to pass ruff
```
Comment (henrikfo) [inline leakpro/input_handler/modality_extensions/image_extension.py:34]: Is it preferable to  "def augmentation(self:Self, data:Tensor, n_aug:int) -> Tensor: # noqa: ARG002" ?

File: leakpro/attacks/minv_attacks/plgmi.py [hunk 3]
```diff
@@ -170,94 +174,84 @@ def prepare_attack(self:Self) -> None:
...
-        # Train the GAN
         if not self.gan_handler.trained_bool:
-            logger.info("GAN not trained, getting psuedo labels")
-            # Top-n-selection to get pseudo labels
+            logger.info("GAN not trained, getting pseudo labels")
             self.pseudo_loader = self.top_n_selection()
+
             logger.info("Training the GAN")
             # TODO: Change this input structure to just pass the attack class
             self.handler.train_gan(pseudo_loader = self.pseudo_loader,
```
Comment (henrikfo) [inline leakpro/attacks/minv_attacks/plgmi.py:194]: checkpoint_interval is removed from handler.train_gan() function call but is still defined in the train_gan() function, celeba_plgmi_handler.py line 96 and the audit.yaml. It is also still used in the train_gan() function; `if i % checkpoint_interval == 0 and i > 0:`

File: leakpro/input_handler/modality_extensions/image_extension.py (hunk 2 above)
Comment (henrikfo) [reply leakpro/input_handler/modality_extensions/image_extension.py:34]: Not particularly important for the approval of the PR.

File: leakpro/attacks/minv_attacks/plgmi.py [hunk 4]
```diff
@@ -170,94 +174,84 @@ def prepare_attack(self:Self) -> None:
...
-
-        # Save the trained generator
+            logger.info("GAN already trained, skipping training.")
 
 
     def run_attack(self:Self) -> MinvResult:
         """Run the attack."""
         logger.info("Running the PLG-MI attack")
         # Define image metrics class
+        # if getattr(self, "reference_model", None) is None:  # noqa: ERA001
```
Comment (henrikfo) [inline leakpro/attacks/minv_attacks/plgmi.py:220]: I don't really understand these comment, is it for a guide of future implementation?

File: leakpro/attacks/minv_attacks/plgmi.py [hunk 5]
```diff
@@ -170,94 +174,84 @@ def prepare_attack(self:Self) -> None:
...
-                   lr: float =2e-2,
-                   iter_times: int = 10) -> torch.tensor:
+                   iter_times: int = 10,
+                   augment: bool = True) -> torch.tensor:
         """Find the optimal latent vectors z for labels y.
 
         Args:
         ----
             y (torch.tensor): The class labels.
             lr (float): The learning rate for optimization.
```
Comment (henrikfo) [inline leakpro/attacks/minv_attacks/plgmi.py:248]: Update arguments in the documentation to match the arguments of the function

File: leakpro/attacks/utils/gan_losses.py [hunk 6]
```diff
@@ -5,24 +5,51 @@
 AVAILABLE_LOSSES = ["hinge", "dcgan"]
 
 
-def max_margin_loss(out: torch.Tensor, iden: torch.Tensor) -> torch.Tensor:
-    """Compute the max margin loss.
+# def max_margin_loss(out: torch.Tensor, iden: torch.Tensor) -> torch.Tensor:
```
Comment (henrikfo) [inline leakpro/attacks/utils/gan_losses.py:8]: Do we need the commented version of the max_margin_loss(...)?

File: examples/minv/celebA/main_celebA_minv.ipynb [hunk 7]
```diff
@@ -58,37 +58,38 @@
    "source": [
```
Comment (henrikfo) [inline examples/minv/celebA/main_celebA_minv.ipynb:58]: <pre class="ql-syntax" spellcheck="false">Line #28.    #y = torch.tensor([0, 1, 2, 3]).to(device) # fixed labels
</pre><p>The duplication and commented "y" is not needed </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aidotse/LeakPro/pull/404/discussion/#comment-3074074998'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

**Review by henrikfo — CHANGES_REQUESTED** [review]
Added a few small comments and feedbacks! Everything but these things looks great to me!  I might have missed some context for them so let me know if they are incorrect, otherwise, with these changes we can merge!
